File: leetcode/strings/ValidPalindrome.java
public class ValidPalindrome {
    
    public static boolean validPalindrome(String s) {
        int i=0;
        int j=s.length()-1;

        while(i<j) {
            if (s.charAt(i) != s.charAt(j)) {
                return isPalindrome(s, i+1, j) || isPalindrome(s, i, j-1);
            }
            i++;
            j--;
        }
        return true;
    }

    public static boolean isPalindrome(String s, int i, int j){
        while (i<j) {
            if (s.charAt(i) != s.charAt(j)) return false;
            i++;
            j--;
        }
        return true;
    }

    public static void main(String[] args) {
        System.out.println(validPalindrome("tebbem"));
    }
}

File: leetcode/strings/two_pointers/valid_palindrome.py
"""
680. Valid Palindrome II
Solved
Easy
Topics
Companies
Given a string s, return true if the s can be palindrome after deleting at most one character from it.
"""


class ValidPalindrome(object):
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """

        def isPalindrome(s, left, right, deletes):
            numDeletesAllowed = 1
            while left < right:
                if s[left] != s[right]:
                    if deletes >= numDeletesAllowed:
                        return False
                    else:
                        return isPalindrome(
                            s, left + 1, right, deletes + 1
                        ) or isPalindrome(s, left, right - 1, deletes + 1)
                else:
                    left += 1
                    right -= 1

            return True

        return isPalindrome(s, 0, len(s) - 1, 0)
